### PR TITLE
hypre_Item

### PR DESCRIPTION
**Pull request by waynemitchell**
Introduces `hypre_Item` in order to further unify code between sycl and cuda/hip implementations and remove `if`/`def`ing. `hypre_Item` should be passed as the first argument to all GPU kernel functions throughout the code. In the case of sycl, `hypre_Item` is defined as `sycl::nd_item<1>`, and this supplies the required item argument for sycl kernels. In the case of cuda/hip, `hypre_Item` is simply defined as `void*` and is ignored inside kernel functions.

Some thread/warp level functions have been modified/added/renamed in order to unify implementation.

Modification and renaming:
`hypre_cuda_get_grid_thread_id` -> `hypre_gpu_get_grid_thread_id`
`hypre_cuda_get_grid_warp_id` -> `hypre_gpu_get_grid_warp_id`
`hypre_cuda_get_lane_id` -> `hypre_gpu_get_grid_thread_id`

Added functions:
`warp_shuffle_sync`
`warp_any_sync`

Questions:
1. Should we call this something more specific than `hypre_Item`?

### CODE REVIEW

File: src/utilities/_hypre_utilities.hpp [hunk 1]
```diff
@@ -778,7 +782,7 @@ hypre_int hypre_cuda_get_grid_num_threads()
 /* return the flattened thread id in grid */
 template <hypre_int bdim, hypre_int gdim>
 static __device__ __forceinline__
-hypre_int hypre_cuda_get_grid_thread_id()
```
Comment (liruipeng) [inline src/utilities/_hypre_utilities.hpp:794]: Why do we need to change the CUDA code here?

Comment (liruipeng) [reply src/utilities/_hypre_utilities.hpp:794]: never mind. I guess I know why.

File: src/utilities/device_utils.c [hunk 2]
```diff
@@ -416,17 +416,9 @@ hypre_GetDefaultDeviceGridDimension( HYPRE_Int n,
 }
 
 __global__ void
-hypreGPUKernel_IVAXPY(
```
Comment (liruipeng) [inline src/utilities/device_utils.c:427]: So, all (CUDA) kernels will have `hypre_Item &item` which is `void *` as the first arg. in the future? It's OK to waste a register space I guess. Maybe the compiler is smart enough to optimize it out.

**Comment by liruipeng** [comment]
I like the changes in this PR to unify sycl with CUDA/HIP. I think it should be merged into master as soon as we can. Thanks @waynemitchell .

**Comment by rfalgout** [comment]
Hi Wayne and Ruipeng,

Just wanted to jump in with a couple comments/questions:

* We probably should use a more specific name than `hypre_Item`.  Do we have any other GPU-related types we define that we can mimic?
* I was originally thinking this would only affect wrappers for the HIP routines that use `item`.  For example, I was envisioning a wrapper that mimics the behavior of `hypre_cuda_get_grid_thread_id<1, 1>()` and `hypre_sycl_get_grid_warp_id(item)`.  If the wrapper were defined as a macro, then the issue with the syntax could be dealt with.  This would also avoid the question from Ruipeng of using a register.  We would probably also need a `hypre_ItemCreate()` routine/macro that would work in tandem with the `hypre_Item` type (here again, I'm being lazy with names).

Does this make any sense or am I missing something?  Maybe we can chat a bit at our meeting today as well.

Thanks @waynemitchell for doing this!

**Comment by rfalgout** [comment]
Hi @waynemitchell and @liruipeng .  I do think we should be more specific with the name of `hypre_Item` since it is very device/GPU specific and defined in `utilities/device_utils.h`.  Maybe `hypre_DeviceItem` or `hypre_GPUItem` or `hypre_WarpItem` or ...?  I don't understand the meaning of this object well enough to suggest a good name, but hopefully you get the idea.

**Review by rfalgout — APPROVED** [review]
This looks good to me modulo the naming question for hypre_Item.  Thanks!

**Comment by waynemitchell** [comment]
> Hi @waynemitchell and @liruipeng . I do think we should be more specific with the name of `hypre_Item` since it is very device/GPU specific and defined in `utilities/device_utils.h`. Maybe `hypre_DeviceItem` or `hypre_GPUItem` or `hypre_WarpItem` or ...? I don't understand the meaning of this object well enough to suggest a good name, but hopefully you get the idea.

I would be fine with either `hypre_DeviceItem` or `hypre_GPUItem`. Thoughts from others?

**Comment by rfalgout** [comment]
> I would be fine with either `hypre_DeviceItem` or `hypre_GPUItem`. Thoughts from others?

Our memory model uses the abstract terminology "Host" and "Device", so `hypre_DeviceItem` makes sense from that perspective.  Is the idea of an "Item" something that we can generalize/extend to an abstract "Device" setting?  If so, I vote for that.

**Comment by liruipeng** [comment]
`hypre_DeviceItem` sounds good to me. Thanks!

File: src/seq_mv/csr_spgemm_device_numer.in [hunk 3]
```diff
@@ -19,7 +19,7 @@
 template HYPRE_Int
 hypre_spgemm_numerical_with_rownnz
 < HYPRE_SPGEMM_BIN, NUMER_HASH_SIZE[HYPRE_SPGEMM_BIN], T_GROUP_SIZE[HYPRE_SPGEMM_BIN], true >
-( HYPRE_Int m, HYPRE_Int *row_ind, HYPRE_Int k, HYPRE_Int n, bool need_ghash,
```
Comment (liruipeng) [inline src/seq_mv/csr_spgemm_device_numer.in]: I might be confused. This is a host function. Why do we need the `item`?

File: src/seq_mv/csr_spgemm_device_symbl.in [hunk 4]
```diff
@@ -19,23 +19,23 @@
 template HYPRE_Int
 hypre_spgemm_symbolic_rownnz
 < HYPRE_SPGEMM_BIN, SYMBL_HASH_SIZE[HYPRE_SPGEMM_BIN], T_GROUP_SIZE[HYPRE_SPGEMM_BIN], true >
-( HYPRE_Int m, HYPRE_Int *row_ind, HYPRE_Int k, HYPRE_Int n, bool need_ghash, HYPRE_Int *d_ia,
```
Comment (liruipeng) [inline src/seq_mv/csr_spgemm_device_symbl.in]: I might be confused. This is a host function. Why do we need the item?

File: src/seq_mv/csr_spgemm_device.h [hunk 5]
```diff
@@ -97,7 +97,8 @@ void group_read(const HYPRE_Int *ptr, bool valid_ptr, HYPRE_Int &v1, HYPRE_Int &
    {
       /* lane = warp_lane
        * Note: use "2" since assume HYPRE_WARP_SIZE divides (blockDim.x * blockDim.y) */
-      const HYPRE_Int lane = hypre_cuda_get_lane_id<2>();
+      hypre_Item item;
+      const HYPRE_Int lane = hypre_gpu_get_lane_id<2>(item);
```
Comment (liruipeng) [inline src/seq_mv/csr_spgemm_device.h:101]: I assume @waynemitchell has implemented the dimension 2 and 3 versions of these functions (`hypre_gpu_get_lane_id`, etc) for sycl.

**Review by liruipeng — APPROVED** [review]
Looks good to me, other than a few  questions I commented. Thank you @waynemitchell !

File: src/seq_mv/csr_spgemm_device_symbl.in (hunk 4 above)
Comment (waynemitchell) [reply src/seq_mv/csr_spgemm_device_symbl.in]: I thought these were device functions for some reason. This is fixed now.

File: src/seq_mv/csr_spgemm_device.h (hunk 5 above)
Comment (waynemitchell) [reply src/seq_mv/csr_spgemm_device.h:101]: I've not implemented this for sycl yet... but I will leave that for future work when I port the rest of the custom spgemm routine. For the current sycl functionality, it is not needed.

**Comment by waynemitchell** [comment]
I've renamed `hypre_Item` to `hypre_DeviceItem`, and regression tests on tux, lassen, and arcticus are all clean. Unless there are further suggestions or comments, I think this is ready to merge.

**Pull request merged**